community/modules/scripts/spack-install/templates/install_spack.tpl

```diff
@@ -87,26 +87,39 @@ spack compiler find --scope site >> ${LOG_FILE} 2>&1
 
 echo "$PREFIX Installing root spack specs..."
 %{for p in PACKAGES ~}
-  spack install ${p} >> ${LOG_FILE} 2>&1
+  spack install ${INSTALL_FLAGS} ${p} >> ${LOG_FILE} 2>&1
   spack clean -s
 %{endfor ~}
 
 echo "$PREFIX Configuring spack environments"
 %{for e in ENVIRONMENTS ~}
-  {
-    spack env create ${e.name};
-    spack env activate ${e.name};
-  } &>> ${LOG_FILE}
+  %{if e.type == "file" ~}
+    {
+      cat << 'EOF' > ${INSTALL_DIR}/spack_env.yaml
+${e.value}
+EOF
+      spack env create ${e.name} ${INSTALL_DIR}/spack_env.yaml
+      rm -f ${INSTALL_DIR}/spack_env.yaml
+      spack env activate ${e.name}
+    } &>> ${LOG_FILE} 2>&1
+  %{endif ~}
 
-  echo "$PREFIX    Configuring spack environment ${e.name}"
-  %{for p in e.packages ~}
-    spack add ${p} >> ${LOG_FILE} 2>&1
-  %{endfor ~}
+  %{if e.type == "packages" ~}
```

Comment:
There should be a [test config](https://github.com/GoogleCloudPlatform/hpc-toolkit/tree/main/tools/validate_configs/test_configs) to cover a lot of this so that at least the basic functionality is being verified in `make tests`. There is already a test config for [spack-cache](https://github.com/GoogleCloudPlatform/hpc-toolkit/blob/main/tools/validate_configs/test_configs/spack-buildcache.yaml), it might make sense to rename that and add different variants of environments to that file.

Reply:
I was working on adding this, and apparently you can't have `list(map(any))` where the map contains keys that are of type strings and lists.

So, this validate fails on the current version of environments, where they take a list of packages. Given this is already in the code, I didn't remove it. I figured after you're back we can discuss this aspect of it. For now, I added a test config for the new environment format.